<div style="border:solid Chocolate 2px; padding: 40px">
    
 <b>Добрый день, Александр!</b>
Меня зовут Ксения Цыганова, и я буду комментировать твою работу и помогать тебе сделать ее сильнее. В Практикуме мы общаемся на «ты», но если тебе это будет некомфортно, пожалуйста, скажи, и мы будем обращаться друг к другу на «вы».
    
Все мои комментарии делятся на три вида:

    
👍 Здесь все отлично, идем дальше! - <font color='green'> зеленый цвет </font>
    
💡 Все хорошо, однако, можно оптимизировать или рассмотреть другие способы (необязательно для исполнения, чтобы пройти дальше) - <font color='gold'>желтый цвет</font>
    
🤔 Необходимо внести исправления, без которых я не смогу принять проект - <font color='red'>красный цвет</font>
    
    
Свои комментарии я буду выделять заголовком «комментарий ревьюера». Пожалуйста, не удаляй их, это поможет сократить время проверки на повторных ревью.Ты также можешь писать свои комментарии к моим и выделять их любым понравившимся способом, отличным от моего. Если у тебя будут возникать вопросы в течение нашей работы, я с радостью на них отвечу.

Цель моего ревью не критика и не возможность дать тебе готовые ответы, это способ привести тебя в нужное русло, чтобы ты смог развить собственные идеи и решения. Твой будущий работодатель будет поступать похожим образом. 
        
   <b>У тебя все получится!</b>

# Выбор локации для скважины

Допустим, вы работаете в добывающей компании «ГлавРосГосНефть». Нужно решить, где бурить новую скважину.

Вам предоставлены пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. Постройте модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализируйте возможную прибыль и риски техникой *Bootstrap.*

Шаги для выбора локации:

- В избранном регионе ищут месторождения, для каждого определяют значения признаков;
- Строят модель и оценивают объём запасов;
- Выбирают месторождения с самым высокими оценками значений. Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

# План работы по проекту

# Цель проекта

В трёх нефтедобывающих регионах (по 10000 скважин в каждом) взяты пробы нефти, измерено качество сырья объём её запасов.
Необходимо построить модель для обозначения региона с наибольшей прибылью, а также проанализировать возможную среднюю прибыль и риски техникой Bootstrap.

# Описание данных
Данные геологоразведки трёх регионов находятся в data sets: geo_data_0.csv, geo_data_1.csv, geo_data_2.csv
id — уникальный идентификатор скважины;
f0, f1, f2 — три признака точек;
product — объём запасов в скважине (тыс. баррелей).



# Условия задания
 - Использовать модель линейной регрессии (остальные — недостаточно предсказуемые).
 - Исследовать 500 точек при геологоразведке, выбрать 200 с наибольшими запасами для расчёта прибыли.
 - Бюджет на разработку скважин в регионе — 10 млрд рублей.
 - Стоимость одного барреля нефти 450 рублей. Объём указан в 1000 баррелей.
 - Выбрать регионы, в которых вероятность убытков меньше 2.5%. Среди регионов, прошедших оценку рисков, выбрать один с наибольшей средней прибылью.



# План проекта
 - Загрузка и подготовка данных
 - Обучение и проверка модели
 - Подготовка к расчёту прибыли
 - Расчёт прибыли
 - Расчёт рисков
 - Общий вывод



<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Хорошо оформленное подробное вступление.

## Загрузка и подготовка данных

1.1 импортируем необходимые библиотеки

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from IPython.display import display
from numpy.random import RandomState
from scipy import stats
from scipy import stats as st
from sklearn.metrics import r2_score
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
    
Все импорты собраны в одном месте, это здорово. Будет понятно, какие библиотеки потребуется установить в случае чего.

1.2 загружаем данные первого data set. Выводим первые 5 строк

In [2]:
data_0 = pd.read_csv('/datasets/geo_data_0.csv')
display(data_0.head())
display(data_0.info()) #выводим информацию о данных методом info()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 
        
Здесь и далее можно использовать конструкцию try/except или библиотеку os, чтобы избежать ошибок при чтении файлов.

In [3]:
print('дубликатов:', sum(data_0.duplicated())) #считаем кол-во дубликатов

дубликатов: 0


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Плюс за анализ на дубликаты.

1.3 загружаем данные второго data set. Выводим первые 5 строк, считаем дубликаты

In [4]:
data_1 = pd.read_csv('/datasets/geo_data_1.csv')
display(data_1.head())
display(data_1.info()) #выводим информацию о данных методом info()
print('дубликатов:', sum(data_1.duplicated())) #считаем кол-во дубликатов

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

дубликатов: 0


1.4 загружаем данные третьего data set. Выводим первые 5 строк, считаем дубликаты

In [5]:
data_2 = pd.read_csv('/datasets/geo_data_2.csv')
display(data_2.head())
display(data_2.info()) #выводим информацию о данных методом info()
print('дубликатов:', sum(data_2.duplicated())) #считаем кол-во дубликатов

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

дубликатов: 0


1.5 столбец id никакой ценности не несет. Мы его удалим

In [6]:
data_0 = data_0.drop('id', axis=1)
data_1 = data_1.drop('id', axis=1)
data_2 = data_2.drop('id', axis=1)



<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Хорошо, что исключаешь столбец с id.

In [7]:
display(data_0.head())
display(data_1.head())
display(data_2.head())

f0        f1        f2     product
0  0.705745 -0.497823  1.221170  105.280062
1  1.334711 -0.340164  4.365080   73.037750
2  1.022732  0.151990  1.419926   85.265647
3 -0.032172  0.139033  2.978566  168.620776
4  1.988431  0.155413  4.751769  154.036647

f0         f1        f2     product
0 -15.001348  -8.276000 -0.005876    3.179103
1  14.272088  -3.475083  0.999183   26.953261
2   6.263187  -5.948386  5.001160  134.766305
3 -13.081196 -11.506057  4.999415  137.945408
4  12.702195  -8.147433  5.004363  134.766305

f0        f1        f2     product
0 -1.146987  0.963328 -0.828965   27.758673
1  0.262778  0.269839 -2.530187   56.069697
2  0.194587  0.289035 -5.586433   62.871910
3  2.236060 -0.553760  0.930038  114.572842
4 -0.515993  1.716266  5.899011  149.600746

# Вывод
 - Данные успешно загрузили. 
 - Дубликатов нет.
 - Типы данных соответствуют описанию.
 - Столбец id пришлось удалить за ненадобностью и бесполезностью

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
    
Хорошо, что делаешь промежуточные выводы.

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 
        
У тебя есть некоторая повторяемость кода, в этом случае можно использовать пользовательские функции, избегать повторяемые фрагменты кода — хорошая практика.
        
Можно также применить визуализацию для анализа, чтобы были видны распределения данных. 
Визуализация важна не только в работе аналитиков, но и DS-специалистов: можно увидеть важные закономерности в данных, проще интерпретировать результаты моделирования. Да и бизнес будет рад, если мы ещё и какие-то инсайты из данных ему принесём. Я бы посоветовала не только распределения визуализировать, но и взаимосвязи в данных: матрицу диаграмм рассеяния и тепловую карту коэффициентов корреляции Пирсона изобразить. В этой задаче у нас планируется использовать модель линейной регрессии, а она чувствительна к проблеме мультиколлинеарности, поэтому тепловая карта может быть очень полезна.

## Обучение и проверка модели

2.1 Разобъем данные на обучающую и валидационную выборки в соотношении 75:25.
    Используем модель линейной регрессии согласно условиям.
    Выведем средний запас сырья и RMSE модели.

In [8]:
target_0 = data_0['product']
features_0= data_0.drop('product', axis=1)
target_1 = data_1['product']
features_1= data_1.drop('product', axis=1)
target_2 = data_2['product']
features_2 = data_2.drop('product', axis=1)
numeric = ['f0', 'f1', 'f2']

#features_train_0, features_valid_0, target_train_0, target_valid_0 = train_test_split(
  # features_0, target_0, test_size=0.25, random_state=12345)

In [9]:
def model_fit_valid(features, target): # создадим функцию, которря делит данные на обучающую и валидационную выборки (75/25),
                 #обучает модель линейной регрессии (согласно условию), предсказывает на валидационной выборке,        
    
    features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=0.25, random_state=42)
    print('Train shape:', features_train.shape)
    print('Valid shape:', features_valid.shape)
    
    model = LinearRegression()
    model.fit(features_train, target_train)
    predict = model.predict(features_valid)

    predict_mean = predict.mean()
    rmse = mean_squared_error(target_valid, predict)**.5

    print('Cредний запас сырья: {:.4f}'.format(predict_mean))
    print('RMSE: {:.4f}'.format(rmse))
    
    return target_valid, predict


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
    
Хорошая функция. Молодец, что не забываешь вычислять квадратный корень из mse.

2.2 Рассмотрим data_0

In [10]:
target_0_valid, predict_0 = model_fit_valid(features_0, target_0)

Train shape: (75000, 3)
Valid shape: (25000, 3)
Cредний запас сырья: 92.3988
RMSE: 37.7566


2.3 Рассмотрим data_1

In [11]:
target_1_valid, predict_1 = model_fit_valid(features_1, target_1)

Train shape: (75000, 3)
Valid shape: (25000, 3)
Cредний запас сырья: 68.7129
RMSE: 0.8903


2.4 Рассмотрим data_2

In [12]:
target_2_valid, predict_2 = model_fit_valid(features_2, target_2)

Train shape: (75000, 3)
Valid shape: (25000, 3)
Cредний запас сырья: 94.7710
RMSE: 40.1459


# Вывод
 - Самый высокий средний запас сырья = 94.8 обнаружен в месторождениях Региона 3. Второй по средним запасам Регион 1 = 92,4, замыкает тройку Регион 2 со средними запасами = 68,7.
 - в тоже время RMSE регионов лидеров также высок (40.1 и 37,8 соответственно). ВОозможна проблема с точностью (качеством) модели регрессии.
 - в Регионе 2 показатель среднего запаса сырья ниже остальных, днако RMSE = 0.89, что указывает на большую точность предсказаний и качестве построенной модели.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Обучение прошло верно. Выводы верные.

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b>
        
Ты заметил, что возможна проблема с точностью. Это тот самый момент, когда была бы полезна визуализация и более глобокое исследование данных. Если визуализировать диаграммы рессеяния, то станет понятнее, почему так получается.

## Подготовка к расчёту прибыли

3.1 Нужно рассчитать минимальный объём сырья для окупаеммости разработки одной скважины. После этого нужно сравнить этот показатель с нашими расчитанными средними запасами сырья в каждом Data (Регионе)

In [13]:
budget = 10_000 # в млн р
samp_wells = 500
best_wells = 200
rev_per_well = 0.45 # доход с 1 тыс баррель в млн. руб
alfa = 0.025 # допустимый коэффициент убытков
bootstrap_samples = 1000
budget_best_wells = budget/best_wells
min_volume = budget_best_wells/rev_per_well
print('мнимальный объем сырья:', min_volume)

мнимальный объем сырья: 111.11111111111111


# Вывод
 - Минимальный необходимы объем сырья удельно на одну скважину = 111,1. 
 - этот показатель превышает наши средние запасы по всем Регионам

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Расчет правильный, выводы верные, молодец.

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b>

Хорошо называть константы в верхнем регистре, чтобы соответствовать стилевым рекомендациям Python-программистов.

Также можно разделять разряды в длинных числах с помощью <code>_</code>: <code>10000000000</code> и <code>10_000_000_000</code> в качестве значения переменной — одно и тоже, но второй вариант проще читать и с ним удобнее работать.         

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> поправил синтаксис, добавил переменную с допустимым процентом убытков и уменьшил большие цифровые значения переменных. Спасибо за подсказку)
</div>

## Расчёт прибыли и рисков 

4.1 создадим функцию для расчета прибыли в лучших скважинах

In [14]:

def income(target_valid, predict, count):
    #revenue = data['product'] * 450000
    pred_sorted = predict.sort_values(ascending=False).head(200) # сортируем 200 лучших скважин по значениям их предсказаний
    targ = target_valid[pred_sorted.index][:count]
    result = rev_per_well * targ.sum()
    return result
predict_0 = pd.Series(predict_0) # перевел из ndarray в Series
predict_1 = pd.Series(predict_1)
predict_2 = pd.Series(predict_2)

predict_0.index = target_0_valid.index # так как индексы не совпали сразу после предказания, я сопоставил индексы и ошибка счезла
predict_1.index = target_1_valid.index
predict_2.index = target_2_valid.index

print('Прибыль 200 лучших скважин:')
print('Регион 1 = {:.2f} млн. руб.'.format(income(target_0_valid, predict_0, best_wells) - budget))
print('Регион 2 = {:.2f} млн. руб.'.format(income(target_1_valid, predict_1, best_wells) - budget))
print('Регион 3 = {:.2f} млн. руб.'.format(income(target_2_valid, predict_2, best_wells) - budget))


Прибыль 200 лучших скважин:
Регион 1 = 3359.14 млн. руб.
Регион 2 = 2415.09 млн. руб.
Регион 3 = 2598.57 млн. руб.


<div class="alert alert-success">
<h2> Комментарий ревьюера №2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
    
Отлично, что заметил момент с индексами. Теперь все верно.

# Вывод
 - самая высокая прибыль в 200 лучших скважинах получилась у Региона 1 =3,3 млрд р.
 - на втором местс регион 3 = 2,6 млрд р
 - на третьем Регион 2= 2,4 млрд р

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
    
Давай посмотрим на функцию: ты делаешь сортировку, основываясь на целевом значении объема сырья. В задании же сказано "Выберите скважины с максимальными значениями предсказаний". 

Здесь стоит подумать, как это сделать. 
 
Мы выбираем определенное количество лучших скважин именно по прогнозу модели,  если бы мы знали сразу реальное число для новых скважин, модель нам не понадобилась бы.
    
Также стоит учитывать, что в итоге у тебя получается не значение прибыли, а значение оборота. Чтобы посчитать прибыль, нужно еще учесть бюджет на разработку.

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> Заменил изначальный data set на предсказания в модели, отсортировал скважины с максимальными значениями предсказаний. Спасибо за замечание. Я не сразу смог разобраться)
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера №2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Ты молодец, что разобрался, не имеет значения сразу/не сразу. Главное, что ты не бросил попыток. А значит, все получится.

4.2 рассчитаем среднюю прибыль для каждого региона
необходимо создать новую функцию для случайного кол-ва скважин (по заданию их должно быть 500)

In [15]:
def revenue(target_valid, predict, count):
    pred_sorted = predict.sort_values(ascending=False) # сортируем скважины по значениям их предсказаний
    targ = target_valid[pred_sorted.index][:count]
    result = rev_per_well * targ.sum()
    return result


<div class="alert alert-warning">
    <h2> Комментарий ревьюера №2<a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b>

Функция revenue дублирует функцию income, заявленную выше.

In [23]:
revenue_0 = []
state = np.random.RandomState(12345)
count_0 = 0

for i in range(bootstrap_samples): # создаем 1000 выборок
    target_subsample_0 = target_0_valid.sample(n = samp_wells, replace=True, random_state=state)
    pred_subsample_0 = predict_0[target_subsample_0.index]
    rev = revenue(target_subsample_0, pred_subsample_0, best_wells) - budget
    revenue_0.append(rev)
    
revenue_0 = pd.Series(revenue_0)

for o in revenue_0:
    if o < 0 :       #проверяем выборку на убыточность, если минусовое значение, то увеличиваем счетчик
        count_0 += 1

    
mean_0 = revenue_0.mean()
print("Средняя прибыль Регион_1:")
print("{:.2f} млн. руб.".format(mean_0))

confidence_interval = st.t.interval(0.95, len(revenue_0)-1, revenue_0.mean(), np.std(revenue_0, ddof=1))  # посчитаем и выведем 95-% доверительный интервал
print("95%-ый доверительный интервал:", confidence_interval)
    
pvalue = 1. * count_0 / bootstrap_samples
if pvalue < alfa:
    print("риск убыточности Региона 1 = {:.2%}, риск маловероятен".format(pvalue))
else:
    print("риск убыточности Региона 1 = {:.2%} и крайне вероятен".format(pvalue))    

 


Средняя прибыль Регион_1:
435.93 млн. руб.
95%-ый доверительный интервал: (-107.1136797748473, 978.9812240521721)
риск убыточности Региона 1 = 6.10% и крайне вероятен


<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Ключевые значения, используемые для расчетов, лучше сохранить в отдельных переменных, как того требует задание.
Здесь это количество исследуемых точек и количество выборок для Bootstrap.

В целом неудачная практика передавать в код числа в явном виде. Если что-то изменится, то придётся всё переделывать вручную: монотонно и можно ошибиться.        
Также не совсем поняла, почему ты вычисляешь 5% квантиль, обьясни, пожалуйста.

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> при расчете средней прибыли Региона один внес ряд изменений:
    1) написал новую функцию для подсчета прибыли в случайном кол-ве скважин 
    1) в цикле заменил data на предсказания
    2) сопоставил индексы
    3) убрал числа, заменил их переменными
    4) 5%-й квантиль показыает какая прибыль будет в скважинах Региона 1 в 95% случаев. Использовал его, чтобы посмотреть насколько может отличаться среднее значения от того, которое будет в 95% случаев. В последствии я отказался от этой метрики, так как в задании не стояло такого условия
    5) не сразу до меня долшло как посчитать риск возможных убытков, но в цикое поставил услове по проверке выборки на минусовую прибыль и сделал счетчик
    6) посчитал среднюю прибыль, доверительный интервал, чтобы понимать насколько в минус сможет уходить прибыль в выборке, высчитал pvalue и сравнил с заданным условием допустимого % убытка = 2,5%
</div>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера №2<a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 
        
Не совсем верное определение 5% квантиля.

In [24]:
revenue_1 = []
count_1 = 0

for e in range(bootstrap_samples): # создаем 1000 выборок
    target_subsample_1 = target_1_valid.sample(n = samp_wells, replace=True, random_state=state)
    pred_subsample_1 = predict_1[target_subsample_1.index]
    rev_1 = revenue(target_subsample_1, pred_subsample_1, best_wells) - budget
    revenue_1.append(rev_1)
    
revenue_1 = pd.Series(revenue_1)

for u in revenue_1:
    if u < 0 :
        count_1 += 1
    
mean_1 = revenue_1.mean()
print("Средняя прибыль Регион_2:")
print("{:.2f} млн. руб.".format(mean_1))

confidence_interval = st.t.interval(0.95, len(revenue_1)-1, revenue_1.mean(), np.std(revenue_1, ddof=1))  # посчитаем и выведем 95-% доверительный интервал
print("95%-ый доверительный интервал:", confidence_interval)
    
pvalue = 1. * count_1 / bootstrap_samples
if pvalue < alfa:
    print("риск убыточности Региона 2 = {:.2%}, риск маловероятен".format(pvalue))
else:
    print("риск убыточности Региона 2 = {:.2%} и крайне вероятен".format(pvalue))    

Средняя прибыль Регион_2:
498.57 млн. руб.
95%-ый доверительный интервал: (78.22561584002028, 918.912417993557)
риск убыточности Региона 2 = 1.00%, риск маловероятен


In [21]:
revenue_2 = []
count_2 = 0


for t in range(bootstrap_samples): # создаем 1000 выборок
    target_subsample_2 = target_2_valid.sample(n = samp_wells, replace=True, random_state=state)
    pred_subsample_2 = predict_2[target_subsample_2.index]
    rev_2 = revenue(target_subsample_2, pred_subsample_2, best_wells) - budget
    revenue_2.append(rev_2)
    
revenue_2 = pd.Series(revenue_2)

for y in revenue_2:
    if y < 0 :
        count_2 += 1

mean_2 = revenue_2.mean()
print("Средняя прибыль Регион_3:")
print("{:.2f} млн. руб.".format(mean_2))

confidence_interval = st.t.interval(0.95, len(revenue_2)-1, revenue_2.mean(), np.std(revenue_2, ddof=1))  # посчитаем и выведем 95-% доверительный интервал
print("95%-ый доверительный интервал:", confidence_interval)
    
pvalue = 1. * count_2 / bootstrap_samples
if pvalue < alfa:
    print("риск убыточности Региона 3 = {:.2%}, риск маловероятен".format(pvalue))
else:
    print("риск убыточности Региона 3 = {:.2%} и крайне вероятен".format(pvalue))




Средняя прибыль Регион_3:
406.93 млн. руб.
95%-ый доверительный интервал: (-131.7784256694049, 945.641665848101)
риск убыточности Региона 3 = 7.00% и крайне вероятен


# Вывод
 - Используя выборку из 500 скважин, потенциальная прибыль уменьшается в среднем с 2,5 - 3 млрд рублей до 0,4 - 0,5 млрд рублей. 
 - Судя по доверительному интервалу Регион 1 и Регион 3 имеют вероятность убытков, поэтому не подходят (вероятность убытков должна быть меньше 2,5 %). 
 - Регион 2 показал минимальную вероятность убытков и поэтому можем считать его как наименее рискованный. У Региона 2 у единственного левая граница 95-% доверительного интервала - положительная. 

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>

Последний пункт вывода вызывает вопросы.
Как мы использовали модель, чтобы получить такой вывод?
Зачем заказчику нужна наша модель?
Стоит поразмышлять над этим, тогда должно стать понятнее, почему алгоритм действий именно такой.

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
    
Не совсем понятно, что за функция predicted_income. Из-за этого ячейка ниже выдает ошибку. 
Просьба перед тем, как сдавать проект, запусти его сначала и до конца, иначе проект может быть возвращен на доработку без проверки.
На этот раз я произвела проверку, потому что не работают буквально последние 3 ячейки, но это скорее исключение.
    
Также получается, что у тебя дважды проходит процедура bootstrap. Код дублируется.
    
Комментарии в функции не соответсвуют использованным ранее в проекте переменным.
    
Ответь, пожалуйста, на несколько вопросов, это поможет убедиться, что проект тебе понятен:
    
1. В чем логика этого и предыдущего разделов
2. В чем отличие между этим и предыдущим разделом.
3. Зачем нам нужна модель машинного обучения.

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> постараюсь ответить так как понимаю задание:
    1) предыдущий раздел считал вероятную прибыть из 200 самых привлекательных скважин и показал хорошие результаты. Значения прибыли для каждого региона (от 2.4 млрд руб. до 3.2 млрд руб). Однако разведка десятков тысяч скважин что повлечет за собой очень большие с вложениями затраты, поэтому принято решение расчитать прибыль на 500 случайных скважинах.
    2) В текущем разделе мы рассчитываем прибыль, процент и процент риска по случайным 500 скважинам, чтобы не давать бизнесу завышенные ожидания по модели от 200 лучших скважин. При случайном выборе скважин (не менее 500) для исследования, расчетная прибыль падает в несколько раз по сравнению с ситуацией, если бы мы проводили разведку для всего объема скважин. Рекомендуем заказчику определить объем вложений для проведения геолоразведки 1-й скважины и включения данного параметра в бизнес-модель.
    3) модель ML необходима для того, чтобы она, обучаясь, предсказывала прибыльность либо убыточность на выборке. Можно, конечно, это делать считая в ручную каждую скважину, но это не целесообразно. 
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера №2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
    
Да, теперь все стало яснее. Молодец, что доработал.

## Общий Вывод
 - Данные успешно загрузили. 
 - Дубликатов нет.
 - Типы данных соответствуют описанию.
 - Столбец id пришлось удалить за ненадобностью и бесполезностью
 - Самый высокий средний запас сырья = 94.8 обнаружен в месторождениях Региона 3. Второй по средним запасам Регион 1 = 92,4,      - Замыкает тройку Регион 2 со средними запасами = 68,7.
 - В тоже время RMSE регионов лидеров также высок (40.1 и 37,8 соответственно). ВОозможна проблема с точностью (качеством) модели регрессии.
 - Минимальный необходимы объем сырья удельно на одну скважину = 111,1. 
 - Этот показатель превышает наши средние запасы по всем Регионам
 - Наибольшая средняя выручка по 200 лучшим скважинам у нас в Регионе 1 = 3,3 млрд р, на втором местс регион 3 = 2,6 млрд р
на третьем Регион 2= 2,4 млрд р
 - Используя выборку из 500 скважин, потенциальная прибыль уменьшается в среднем с 2,5 - 3 млрд рублей до 0,4 - 0,5 млрд рублей.
 - Судя по доверительному интервалу Регион 1 и Регион 3 имеют вероятность убытков, поэтому не подходят (вероятность убытков должна быть меньше 2,5 %).
 - Регион 2 показал минимальную вероятность убытков и поэтому можем считать его как наименее рискованный. У Региона 2 у единственного левая граница 95-% доверительного интервала - положительная.
 # В соответствии с задачей рекомендую Регион 2 для разработки.

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
    
Когда произведешь перерасчет, необходимо будет также поправить все промежуточные выводы и общий вывод, исходя из новых данных.

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 
    
Если есть желание углубиться в тему, то можно почитать, например, что-то из этих материалов:

"A Gentle Introduction to the Bootstrap Method" (на английском)
https://machinelearningmastery.com/a-gentle-introduction-to-the-bootstrap-method/
        
"Как устроено A/B-тестирование в Авито" (на русском)
https://habr.com/ru/company/avito/blog/454164/
        

<div class="alert alert-success">
<h2> Комментарий ревьюера №2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
    
Хороший общий вывод.

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [ ]  Весь код выполняется без ошибок
- [ ]  Ячейки с кодом расположены в порядке исполнения
- [ ]  Выполнен шаг 1: данные подготовлены
- [ ]  Выполнен шаг 2: модели обучены и проверены
    - [ ]  Данные корректно разбиты на обучающую и валидационную выборки
    - [ ]  Модели обучены, предсказания сделаны
    - [ ]  Предсказания и правильные ответы на валидационной выборке сохранены
    - [ ]  На экране напечатаны результаты
    - [ ]  Сделаны выводы
- [ ]  Выполнен шаг 3: проведена подготовка к расчёту прибыли
    - [ ]  Для всех ключевых значений созданы константы Python
    - [ ]  Посчитано минимальное среднее количество продукта в месторождениях региона, достаточное для разработки
    - [ ]  По предыдущему пункту сделаны выводы
    - [ ]  Написана функция расчёта прибыли
- [ ]  Выполнен шаг 4: посчитаны риски и прибыль
    - [ ]  Проведена процедура *Bootstrap*
    - [ ]  Все параметры бутстрепа соответствуют условию
    - [ ]  Найдены все нужные величины
    - [ ]  Предложен регион для разработки месторождения
    - [ ]  Выбор региона обоснован

<div style="border:solid Chocolate 2px; padding: 40px">
    
<b>Общий вывод по проекту</b>

Александр, благодарю тебя за работу! Проект продуманный, доведенный до конца, для этого нужно много сил, времени и мотивации. Ты все это инвестировал, видно, что ты старался. В проекте много хороших моментов. Однако есть места, которые стоит проработать еще раз, чтобы сделать его сильнее для портфолио.

**Положительные моменты проекта, которые хочу еще раз подчеркнуть:**

1. Выдержана структура проекта,
2. Проведена предварительная работа с данными,
3. Верно обучены модели линейной регрессии для всех трех регионов,
4. Верно произведен расчет минимального обьема сырья,
5. Реализована процедура bootstrap.

**Замечания, на которые стоит обратить внимание при будущей работе:**

1. Возможно, использовать визуализацию для предварительного анализа данных
2. Необходимо исправить функцию расчета прибыли
3. Произвести перерасчет и исправить выводы, исходя из новых значений.
4. Избавиться от дублирующего кода.

Остальные комментарии ты найдешь выше.

**Желаю удачи и жду твой проект на повторное ревью!**



<div style="border:solid Chocolate 2px; padding: 40px">
    
<b>Общий вывод по проекту №2</b>
    
Александр, ты хорошо поработал, теперь все стало на свои места в проекте. Молодец!
Я отправлю еще раз работу тебе назад. На случай, если ты хочешь задать вопросы, может, что-то еще добавить или подправить.
Если нет, то на следующей итерации я приму проект.
    
Желаю успехов!

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> Ксения, спасибо за качественное ревью. Ты мне очень помогла разобраться с тем, с чем в теории изначально я недопонял. С удовольствием приступлю к Сборному проекту и буду рад, если ты снова будешь моим ревьюером)
</div>